#14: Import with certificates (#15)
exasol/exasol-java-tutorial: tls-tutorial/doc/use_your_own_certificate.md

* [MySQL docker image](https://hub.docker.com/_/mysql/)

## Installing the Software

1. Update the machine you are using
    ```shell
    sudo apt update
    sudo apt upgrade
    ```
1. Install Docker
   ```shell
   sudo apt install docker.io
   ```
1. Install Open SSL.
    Note that on Ubuntu OpenSSL should be preinstalled. If not run:
    ```shell
    sudo apt install openssl
    ```
2. Create a directory where we will keep the files that we are creating in the course of this project.
    ```shell
    tutorial_dir="$HOME/tutorials/tls_with_exasol/import"
    mkdir -p "$tutorial_dir"
    cd "$tutorial_dir"
    ```

## Creating and Installing the Certificates

In this part of the tutorial we will create the required TLS certificates and install them on Exasol and MySQL.

### Creating Your own Certification Agency (CA)

To get that right out of the way, yes, in a normal organization, you already have a Certification Agency (CA). But for the purpose of this tutorial, let's pretend you don't.

That means we will create our own, and that in turn gives us the opportunity to create the server certificates we are going to use quickly.

> The script below takes you through the following steps:

Reviewer: There is no script, instead a couple of shell commands to copy.
```suggestion
Do this to create a CA certificate:
```

If you are curious, you can check the modulus of the server key like this:

```shell
openssl rsa -in server.key -text
```

When you do this on your machine you will find that the output matches the modulus in the CSR.

```
RSA Private-Key: (4096 bit, 2 primes)
modulus:
    00:99:af:[...]
[...]
```

The most important puzzle piece in the signing request is that it contains the public key of the requester. In our case the server public key. Without that users of the certificate would not be able to verify the authenticity of the TLS connection.

You can export the public key from the signing request like this:

```shell
openssl req -in server.csr -pubkey -noout
```

If you compare it to the public key contained in the server key pair, you will find it is identical.

```shell
openssl rsa -in server.key -pubout
```

### Singing the Server Certificate

That is something that the organization owning the server would normally send to the certification agency. The request contains everything the CA needs to create a signed certificate.

Since in our tutorial you own both the server and the CA, you can play both parts.

> We earlier briefly already touched the topic of certificate extensions. To define which extensions you want on the server certificate signed by the CA, you need to create a small [configuration file](#x509v3config).

Reviewer: ```suggestion
We earlier briefly already touched the topic of certificate extensions.To define which extensions you want on the server certificate signed by the CA, this creates a small [configuration file](#x509v3config).
```

We keep things simple here though, since this is just a tutorial meant to convey the general idea of server certificate validation.

## Setting up MySQL

1. Install MySQL client and server
   ```shell
   sudo apt install mysql-client mysql-server
   ```
2. Log into MySQL as MySQL root user (on Ubuntu now requires `sudo`)
   ```shell
   sudo mysql -u root
   ```
3. Create a database `shapes` and a user `tutorial_user` for our import experiment
   ```sql
   create database shapes;
   create user 'tutorial_user' identified by 'tutorial';
   grant all on shapes.* to tutorial_user;
   exit
   ```
4. Login as a `tutorial_user`
   ```sql
   mysql -u tutorial_user -p
   ```
5. Populate the database
   ```sql
   create table shapes.shapes (name VARCHAR(40), corners int);
   insert into shapes.shapes values ('point', 1), ('line', 2), ('triangle', 3), ('rectangle', 4);
   exit
   ```

As you can tell we won't win any price for most original database usage here, but that is not the point of the tutorial.

### Installing Certificates in MySQL

> By default, the MySQL server on Ubuntu comes with a self-signed certificate. Our next job will be to use the certificates we previously created.

Reviewer: What’s wrong with that self-signed certificate? We should mention a good reason why we change that.

Author: Added explanation.

1. Install the CA certificate centrally:
   ```shell
   sudo cp ca.crt /usr/local/share/ca-certificates/exasol_tutorial_ca.crt
   sudo update-ca-certificates
   ```
2. Verify that it was installed:
   ```shell
   ls -l /etc/ssl/certs/exasol*
   ```
   Must output something like:
   ```
   lrwxrwxrwx 1 root root 55 Jan 12 11:43 /etc/ssl/certs/exasol_tutorial_ca.pem -> /usr/local/share/ca-certificates/exasol_tutorial_ca.crt
   ```
   Note how the symbolic link changes the file suffice from `.crt` to `.pem`
3. Install the server certificate and key
   ```shell
   sudo mkdir -p /etc/mysql/certs
   sudo cp server.crt /etc/mysql/certs/
   sudo cp server.key /etc/mysql/certs/
   sudo chown mysql:mysql /etc/mysql/certs/server.key
   ```

Note that we need `sudo` privileges here, since we need to install the certificates in a central locations that belong to the `root` user.

> Additionally, we give the private key to the MySQL user. We don't want anyone else to be able to read it, so weakening permissions on that file is not an option.

Reviewer: Should be mentioned that this user is automatically created upon install of MySQL.

Author: Done.

You can check ownership and permissions like this:

```shell
ls -l /etc/mysql/certs/
```

This should list two files:

```
-rw-r--r-- 1 root  root  1919 Jan 12 11:56 server.crt
-rw------- 1 mysql mysql 3326 Jan 12 11:56 server.key
```

Now edit the `[mysqld]` section of the MySQL configuration file and *add* the following lines as `root` user:

```
[mysqld]
ssl
ssl-ca=/etc/ssl/certs/exasol_tutorial_ca.pem
ssl-cert=/etc/mysql/certs/server.crt

> ssl-key=/etc/mysql/certs/server_key.pem

Reviewer: ```suggestion
ssl-key=/etc/mysql/certs/server.key
```

> Now edit the `[mysqld]` section of the MySQL configuration file and *add* the following lines as `root` user:
> 
> ```
> [mysqld]
> ssl
> ssl-ca=/etc/ssl/certs/exasol_tutorial_ca.pem
> ssl-cert=/etc/mysql/certs/server.crt
> ssl-key=/etc/mysql/certs/server_key.pem
> ```
> 
> Also, we need to enable Exasol later to access the MySQL server. Since that will not be access from `localhost` you need to accept all interfaces.
> 
> ```
> bind-address = 0.0.0.0

Reviewer: ```suggestion
Now edit the `[mysqld]` section of the MySQL configuration file /etc/mysql/mysql.conf.d/mysqld.cnf as `root` user:

Add
```
[mysqld]
ssl
ssl-ca=/etc/ssl/certs/exasol_tutorial_ca.pem
ssl-cert=/etc/mysql/certs/server.crt
ssl-key=/etc/mysql/certs/server_key.pem
```

Also, we need to enable Exasol later to access the MySQL server. Since that will not be access from `localhost` you need to accept all interfaces.

Change
```
bind-address = 0.0.0.0
```

Finally, you can connect the MySQL client to the server using TLS:

```shell
mysql -u tutorial_user -p --ssl-mode=REQUIRED -u tutorial_user
```

The source side of our import via TLS is now set up and ready to use.

## Setting up Docker

For the next steps we need a Docker setup

1. Install Docker
    ```shell
    sudo apt install docker.io
    ```
2. Add your current user to the `docker` group
    ```shell
    sudo usermod -aG docker "$USER"
    ```
3. Log out and back in to apply the group change
4. Verify that your current user is in the `docker` group
    ```shell
   groups
    ```

## Installing Exasol via Docker

Exasol's [docker-db](https://github.com/exasol/docker-db) provides a Docker container that is quite convenient for experiments like our tutorial.

The following docker command downloads the container's layers and afterwards starts the container. Expect a couple of minutes until the download and database start are done.

> 

Reviewer: First you need to start the docker daemon with something like sudo dockerd 
We need to explain how the Exasol docker container can be safely shutdown and restarted in case people don’t do the whole tutorial in one go.

Author: Added:
```shell
sudo systemctl enable docker.service
```

Author: Also added a very brief Docker 101.

```shell
docker run --name exasoldb -p 8563:8563 -p 2580:2580 --detach --privileged --stop-timeout 120  exasol/docker-db:7.1.17
```

There are two ports forwarded to the host:

- 8563: database port
- 2580: [BucketFS] port

### Installing the MySQL JDBC driver in Exasol

[Import](https://docs.exasol.com/db/latest/sql/import.htm) into Exasol requires that the database driver of the source database (MySQl in our case) is installed in Exasol.

Installation in the Docker variant requires uploading the driver to a [BucketFS] bucket and updating a configuration file.

1. Download the [MySQL Connector] (JDBC driver)
   ```shell
   wget https://dev.mysql.com/get/Downloads/Connector-J/mysql-connector-j-8.0.31.tar.gz
   ```
2. Get the ID of the Exasol Docker container
   ```shell

>    container_id=$(docker ps -a | grep exasol | sed -e's/ .*//')

Reviewer: Didn’t work for me. I had to do docker ps -a  to find the container ID.

   ```
3. Get the write-password of the default Bucket in BucketFS
   ```shell
   write_pwd=$(docker exec -it "$container_id" cat /exa/etc/EXAConf | grep WritePasswd | sed -e's/^.* = //' -e's/[\n\r]//g' | base64 --decode)
   ```
4. Copy the driver to the default Bucket
   ```shell
   curl -vX PUT -T mysql-connector-j-8.0.31/mysql-connector-j-8.0.31.jar "http://w:$write_pwd@localhost:2580/default/drivers/jdbc/mysql-connector-j-8.0.31.jar"
   ```
5. Create a driver configuration (`settings.cfg`)
   ```shell
   echo 'DRIVERNAME=MYSQL_JDBC
   JAR=mysql-connector-j-8.0.31.jar
   DRIVERMAIN=com.mysql.cj.jdbc.Driver
   PREFIX=jdbc:mysql:
   NOSECURITY=YES
   FETCHSIZE=100000
   INSERTSIZE=-1' > settings.cfg

>    ```

Reviewer: Make sure that file has no leading spaces.

Author: That is a pretty good point. While in the rendered version there are no leading spaces, people get them when copying from the source.

Author: Added disclaimer.

6. Upload the configuration to `drivers/jdbc` in the default bucket:
   ```shell
   curl -vX PUT -T settings.cfg "http://w:$write_pwd@localhost:2580/default/drivers/jdbc/settings.cfg"
   ```
7. Verify this installation:
   ```shell
   curl http://localhost:2580/default
   ```
   Must produce a listing that looks similar to this:
   ```
   EXAClusterOS/ScriptLanguages-standard-EXASOL-7.1.0-slc-v4.0.0-CM4RWW6R.tar.gz
   drivers/jdbc/mysql-connector-j-8.0.31.jar
   drivers/jdbc/settings.cfg
   ```
   
### Copying the Certificate to Exasol

**Information:** this part will be added in a separate pull request.

## Run the Import

In this last part of the tutorial, we import the data from the `shapes` database and `shapes` table in MySQl into Exasol.

One last piece of preparation is that you get the IP address of the Docker host (i.e. the Ubuntu machine you run the tutorial on).

```shell
ip address show docker0 | grep inet
```

Please, note down the IP address. You will need it for the `IMPORT` command.

> 

Reviewer: Isn’t that always 172.17.0.1 ?

Author: Unfortunately not. It depends on the setup of the Docker host.

   ```shell
   wget https://github.com/xo/usql/releases/download/v0.13.5/usql_static-0.13.5-linux-amd64.tar.bz2
   tar --bzip2 -xvf usql_static-0.13.5-linux-amd64.tar.bz2
   ```

Next, connect to the Exasol server

```shell
./usql_static exa://sys:exasol@localhost?validateservercertificate=0
```

Since we forwarded the default port, we can directly connect to `localhost`, even if the Exasol database is running in a Docker container.

You are probably asking yourself why we skip certificate validation here. The reason is that we otherwise would have to outfit Exasol with a server certificate too and that is beyond the scope of this particular tutorial for now (later versions might add this).

In order to run the import, we need a target table on Exasol, so let's prepare it.

```sql
CREATE SCHEMA target_schema;
CREATE TABLE target_schema.shapes_target(name VARCHAR(40), corners INT);
```

You are now set up and ready to run the import.

> ```shell

Reviewer: ```suggestion
```sql
```

IMPORT INTO target_schema.shapes_target
FROM JDBC AT 'jdbc:mysql://<ip-address-of-docker-host>/shapes?sslMode=REQUIRED'
USER 'tutorial_user' IDENTIFIED BY 'tutorial'

> STATEMENT 'SELECT * FROM shapes';

Reviewer: ```suggestion
STATEMENT 'SELECT * FROM shapes.shapes';
```

[Subject key identifier](https://www.rfc-editor.org/rfc/rfc3280#section-4.2.1.2) and authority key identifier are unique identifiers that are used in certificate chaining. Note that in our root CA example they are identical.

The `critical` flag after an extension definition means that any system using the certificate _must_ know and process it. The system can ignore extensions that are not marked like this. Case in point: the key identifiers are non-critical because there is a fallback in the form of issuer and subject, although the IDs are obviously the better alternative for machine evaluation.

The `CA:TRUE` entry is pretty self-explaining. You are looking at a CA certificate.

Finally, there is a block that contains the signature of the certificate, which in this case is not particularly useful, since it is self-signed. Since that is the beginning of the chain, and it contains the public key that is required to validate the signature, you have the dreaded chicken-of-the-egg problem of all IT security here. That means the *root* CA certificates must be installed on a client by means of a trusted channel. 

> ### Creating a Singing Request for a Server Certificate

Reviewer: ```suggestion
### Creating a Signing Request for a Server Certificate
```

Note this in this step you will create a certificate singing request (CSR) (see [RFC2986](#rfc2986)) instead of an actual certificate. Think of the signing request a precursor of a signed certificate.

1. Create a key pair _without_ passphrase for the server:
    ```shell
    openssl genrsa -out server.key 4096
    ```
2. You will be prompted for a password. Use `tutorial` for this demonstration.
   In a real world scenario you would obviously choose a secure, non-guessable password.
3. Create a singing request
   ```shell
   openssl req -new -nodes -key server.key -sha256 -out server.csr -subj '/CN=TLS Tutorial MySQL Server/C=DE/L=Bavaria/O=Tutorial Organization'
    ```

Note how the `-x509` switch is gone from the command. This time we really want a signing request, not a self-signed certificate.

If you look real close, you might wonder why there is no `-days` switch here. The reason is simple: the CA decides how long a certificate it creates is valid. Not the requester.

> ### Reading a Singing Request

Reviewer: ```suggestion
### Reading a Signing Request
```

> Let's take a look at the generated singing request to get a better understanding of its contents.

Reviewer: ```suggestion
Let's take a look at the generated signing request to get a better understanding of its contents.
```

    Signature Algorithm: sha256WithRSAEncryption
         8c:33:[...]
```

As you can see, there is not much to it. It looks like a half-finished certificate. Just a text representation of the data you provided on the commandline.

The modulus is a property that the private and public key of a server share. Its purpose is allowing to verify key authenticity.

If you are curious, you can check the modulus of the server key like this:

```shell
openssl rsa -in server.key -text
```

When you do this on your machine you will find that the output matches the modulus in the CSR.

```
RSA Private-Key: (4096 bit, 2 primes)
modulus:
    00:99:af:[...]
[...]
```

The most important puzzle piece in the signing request is that it contains the public key of the requester. In our case the server public key. Without that users of the certificate would not be able to verify the authenticity of the TLS connection.

You can export the public key from the signing request like this:

```shell
openssl req -in server.csr -pubkey -noout
```

If you compare it to the public key contained in the server key pair, you will find it is identical.

```shell
openssl rsa -in server.key -pubout
```

> ### Singing the Server Certificate

Reviewer: ```suggestion
### Signing the Server Certificate
```

   ```
   Note how the symbolic link changes the file suffice from `.crt` to `.pem`
3. Install the server certificate and key
   ```shell
   sudo mkdir -p /etc/mysql/certs
   sudo cp server.crt /etc/mysql/certs/
   sudo cp server.key /etc/mysql/certs/
   sudo chown mysql:mysql /etc/mysql/certs/server.key
   ```

Note that we need `sudo` privileges here, since we need to install the certificates in a central locations that belong to the `root` user.

Additionally, we give the private key to the MySQL user on OS level. This user is aptly called `mysql`, belongs to a group of the same name and is created automatically when you install MySQL on Ubuntu. We don't want anyone else to be able to read it, so weakening permissions on that file is not an option.

You can check ownership and permissions like this:

```shell
ls -l /etc/mysql/certs/
```

This should list two files:

```
-rw-r--r-- 1 root  root  1919 Jan 12 11:56 server.crt
-rw------- 1 mysql mysql 3326 Jan 12 11:56 server.key
```

Now edit the `[mysqld]` section of the MySQL configuration file `/etc/mysql/mysql.conf.d/mysqld.cnf `as `root` user:

Add

> ```

Reviewer: The content of the config file that should be added/changed is missing

Author: Something strange happened here with taking over the suggestions. Added the missing parts.

The most important part here are the cipher and the ticket lifetime of 5 minutes.

**Side-note:** if you ever experience TLS connections breaking of after a certain time, you should check first this session ticket lifetime and then any intermediate network components like firewalls. Sometimes they interfere with the connection renewal process.

Finally, you can connect the MySQL client to the server using TLS:

```shell
mysql -u tutorial_user -p --ssl-mode=REQUIRED -u tutorial_user
```

The source side of our import via TLS is now set up and ready to use.

## Setting up Docker

For the next steps we need a Docker setup

1. Install Docker
   ```shell
   sudo apt install docker.io
   ```
2. Add your current user to the `docker` group
   ```shell
   sudo usermod -aG docker "$USER"
   ```
3. Log out and back in to apply the group change
4. Verify that your current user is in the `docker` group
   ```shell
   groups
   ```
5. Tell the system control to auto-start the docker daemon:
   ```shell
   sudo systemctl enable docker.service
   ```
6. Start the docker daemon without restaring the machine:
   ```shell

>    sudo systemctl enable docker.service

Reviewer: same cmd twice

Author: Right that should have been `start`. Fixed.

   ```
   Please make sure to copy this from the rendered version (not the Markdown source) of this tutorial. You want **no leading spaces** in the configuration file.
6. Upload the configuration to `drivers/jdbc` in the default bucket:
   ```shell
   curl -vX PUT -T settings.cfg "http://w:$write_pwd@localhost:2580/default/drivers/jdbc/settings.cfg"
   ```
7. Verify this installation:
   ```shell
   curl http://localhost:2580/default
   ```
   Must produce a listing that looks similar to this:
   ```
   EXAClusterOS/ScriptLanguages-standard-EXASOL-7.1.0-slc-v4.0.0-CM4RWW6R.tar.gz
   drivers/jdbc/mysql-connector-j-8.0.31.jar
   drivers/jdbc/settings.cfg
   ```
   
### Copying the Certificate to Exasol

**Information:** this part will be added in a separate pull request.

## Run the Import

In this last part of the tutorial, we import the data from the `shapes` database and `shapes` table in MySQl into Exasol.

One last piece of preparation is that you get the IP address of the Docker host (i.e. the Ubuntu machine you run the tutorial on).

```shell
ip address show docker0 | grep inet
```

Please, note down the IP address. You will need it for the `IMPORT` command.

> We need a why to talk to Exasol, so please install the [usql](https://github.com/xo/usql) commandline SQL client on the Ubuntu machine. Current versions come with the [Exasol Go driver preinstalled](https://github.com/xo/usql#supported-database-schemes-and-aliases).

Reviewer: ```suggestion
We need a way to talk to Exasol, so please install the [usql](https://github.com/xo/usql) commandline SQL client on the Ubuntu machine. Current versions come with the [Exasol Go driver preinstalled](https://github.com/xo/usql#supported-database-schemes-and-aliases).
```

3. Install the server certificate and key
   ```shell
   sudo mkdir -p /etc/mysql/certs
   sudo cp server.crt /etc/mysql/certs/
   sudo cp server.key /etc/mysql/certs/
   sudo chown mysql:mysql /etc/mysql/certs/server.key
   ```

Note that we need `sudo` privileges here, since we need to install the certificates in a central locations that belong to the `root` user.

Additionally, we give the private key to the MySQL user on OS level. This user is aptly called `mysql`, belongs to a group of the same name and is created automatically when you install MySQL on Ubuntu. We don't want anyone else to be able to read it, so weakening permissions on that file is not an option.

You can check ownership and permissions like this:

```shell
ls -l /etc/mysql/certs/
```

This should list two files:

```
-rw-r--r-- 1 root  root  1919 Jan 12 11:56 server.crt
-rw------- 1 mysql mysql 3326 Jan 12 11:56 server.key
```

Now edit the `[mysqld]` section of the MySQL configuration file `/etc/mysql/mysql.conf.d/mysqld.cnf `as `root` user and add:

```
[mysqld]
ssl
ssl-ca=/etc/ssl/certs/exasol_tutorial_ca.pem
ssl-cert=/etc/mysql/certs/server.crt
ssl-key=/etc/mysql/certs/server.key
```

> Also, we need to enable Exasol later to access the MySQL server. Since that will not be access from `localhost` you need to accept all interfaces.

Reviewer: ```suggestion
Also, we need to enable Exasol later to access the MySQL server. Since that access will not be from `localhost`, you need to accept all interfaces.
```
